File: Xenon/Xenon/Spore.ts
import CBullet = require("Bullet");
import CShip = require("Ship");
import gsCVector = require("Vector");
import enums = require("Enums");
import gsCControls = require("Controls");
import gsCTimer = require("Timer");
import Pickup = require("Pickup");
import CSporeGenerator = require("SporeGenerator");

class CSpore extends CBullet {

    private SPORE_HOMING_DELAY: number = 0.5;
    private m_delay_timer: gsCTimer;
    private m_killed_by_player: boolean;

    constructor() {
        super();
        this.m_name = "SPORE";
    }

    //-------------------------------------------------------------

    public activate(): boolean {
        if (!this.isActive()) {
            this.m_timer.start();
            this.m_delay_timer.start();
        }
        return super.activate();
    }

    //-------------------------------------------------------------

    public update(controls: gsCControls, gameTime: gsCTimer): boolean {
        this.gameTime = gameTime;
        if (this.m_shield == 0) {
            //super.explode();
            super.kill();
            return true;
        }

        //if (m_delay_timer.getState() != gsTIMER_ACTIVE || m_delay_timer.getTime() >= SPORE_HOMING_DELAY)
        {
            //this.m_delay_timer.reset();
            var ship: CShip = this.m_scene.findShip();
            if (ship != null) {
                //var rel_pos: gsCVector = ship.getPosition(). - this.m_position;
                //rel_pos.Normalize();
                //this.m_velocity = rel_pos * this.getActorInfo().m_speed[this.m_grade];
            }
        }

        this.m_position.plusEquals(this.m_velocity);
        super.animate(enums.AnimationMode.ANIMATE_LOOP);
        return true;
    }

    //-------------------------------------------------------------

    public onKilled(): void {
        this.m_killed_by_player = true;
        super.onKilled();
    }

    //-------------------------------------------------------------

    public kill(): void {
        if (this.getOwner() != null && (<CSporeGenerator>this.getOwner()).sporeKilled(this.m_killed_by_player)) {
            var s: Pickup.CScorePickup = new Pickup.CScorePickup();
            this.m_scene.addActor(s);
            s.setPosition(this.getPosition());
            s.activate();
        }
        super.kill();
    }

    //-------------------------------------------------------------

    public onLeavingScreen(): void {
    }

    //-------------------------------------------------------------

}
export = CSpore;

File: Xenon-Original_C++_Code/xenon/source/spore.cpp
//-------------------------------------------------------------
//
// Class:	CSpore
//
// Author:	John M Phillips
//
// Started:	06/05/00
//
// Base:	CBullet
//
// Derived:	None
//
//-------------------------------------------------------------

#include "game.h"

//-------------------------------------------------------------

CSpore::CSpore()
{
	m_killed_by_player = false;
}

//-------------------------------------------------------------

CSpore::~CSpore()
{
}

//-------------------------------------------------------------

bool CSpore::activate()
{
	if (!isActive()) {
		m_timer.start();
		m_delay_timer.start();
		}

	return CActor::activate();
}

//-------------------------------------------------------------

bool CSpore::update(Controls *controls)
{
	if (m_shield == 0) {
		explode();
		kill();
		return true;
		}

	if (m_delay_timer.getState() != gsTIMER_ACTIVE ||
		m_delay_timer.getTime() >= SPORE_HOMING_DELAY) {
		m_delay_timer.reset();

		CShip *ship = m_scene->findShip();

		if (ship) {
			gsCVector rel_pos = ship->getPosition() - m_position;

			rel_pos.normalize();

			m_velocity = rel_pos * getActorInfo().m_speed[m_grade];
			}
		}

	m_position += m_velocity;

	animate(ANIMATE_LOOP);

	return true;
}

//-------------------------------------------------------------

void CSpore::onKilled()
{
	m_killed_by_player = true;

	CActor::onKilled();
}

//-------------------------------------------------------------

void CSpore::kill()
{
	if (getOwner() &&
		((CSporeGenerator *) getOwner())->sporeKilled(m_killed_by_player)) {
		CScorePickup *s = new CScorePickup;
		m_scene->addActor(s);
		s->setPosition(getPosition());
		s->activate();
		}

	CActor::kill();
}

//-------------------------------------------------------------

void CSpore::onLeavingScreen()
{
}

//-------------------------------------------------------------
